segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/ConcurrencyManager.java

```diff
@@ -0,0 +1,225 @@
+     * ** Exception is if both throughput and the fill ratio are declining, in which case nothing is done.
+     *
+     * @return The degree of parallelism, whether it was updated or not.
+     */
+    int getOrUpdateParallelism() {
+        if (this.minParallelism == this.maxParallelism) {
+            // Min == Max, so the parallelism will never change.
+            return getCurrentParallelism();
+        }
+
+        final long time = this.timeSupplier.get();
+        synchronized (this.snapshotLock) {
+            // Calculate the most recent throughput and Fill Ratio (of those items that have just been written).
+            final long elapsedMillis = (time - this.lastSnapshot.timeStamp) / AbstractTimer.NANOS_TO_MILLIS;
+            if (elapsedMillis >= STALE_MILLIS) {
+                // Too long since the last update; the data we have is no longer relevant. Reset.
+                resetSnapshot(time);
+            } else if (elapsedMillis >= UPDATE_PERIOD_MILLIS) {
+                // Enough time elapsed since the last update; update the snapshot.
+                updateSnapshot(time, elapsedMillis);
+            }
+
+            return this.lastSnapshot.parallelism;
+        }
+    }
+
+    /**
+     * Updates the current snapshot based on the current data.
+     *
+     * @param time          The current time, in millis.
+     * @param elapsedMillis The elapsed time, in millis, since the last time this method ran.
+     */
+    @GuardedBy("snapshotLock")
+    private void updateSnapshot(long time, long elapsedMillis) {
+        final double recentThroughput;
+        final long recentLatency;
+        synchronized (this) {
+            recentThroughput = (double) this.recentTotalWrittenLength / elapsedMillis;
+            recentLatency = this.recentWriteCount == 0 ? 0 : this.recentTotalLatencyMillis / this.recentWriteCount;
+            this.recentWriteCount = 0;
+            this.recentTotalWrittenLength = 0;
+            this.recentTotalLatencyMillis = 0;
+        }
+
+        final double throughputDifference = recentThroughput - this.lastSnapshot.throughput;
+        final long latencyDifference = recentLatency - this.lastSnapshot.latency;
+        boolean significantThroughputChange = Math.abs(throughputDifference / this.lastSnapshot.throughput) >= SIGNIFICANT_DIFFERENCE;
+        boolean significantLatencyChange = this.lastSnapshot.latency == 0
+                ? latencyDifference != 0
+                : Math.abs((double) latencyDifference / this.lastSnapshot.latency) >= SIGNIFICANT_DIFFERENCE;
+
+        int delta = 0;
+        if (significantThroughputChange) {
+            // Throughput changed significantly. Adjust parallelism in the same direction as throughput change.
+            delta = (int) Math.signum(throughputDifference);
+            if (throughputDifference < 0 && significantLatencyChange && recentLatency < this.lastSnapshot.latency && this.lastSnapshot.delta < 0) {
```

Comment:
If throughput went down and latency went up after we decreased the parallelism, what do we want to do?

Reply:
Good point. Fixed it.